File: pact-jvm-model/src/main/scala/au/com/dius/pact/model/PactMerge.scala
package au.com.dius.pact.model

import au.com.dius.pact.model.v3.messaging.MessagePact

import scala.collection.JavaConversions

trait MergeResult

case class MergeSuccess(result: Pact) extends MergeResult
case class MergeConflict(result: Seq[(Interaction, Interaction)]) extends MergeResult

object PactMerge {
  def merge(newPact: Pact, existing: Pact) : MergeResult = {
    val interactions = JavaConversions.collectionAsScalaIterable(existing.getInteractions).toSeq
    val otherInteractions = JavaConversions.collectionAsScalaIterable(newPact.getInteractions).toSeq
    val failures: Seq[(Interaction, Interaction)] = for {
      a <- interactions
      b <- otherInteractions
      if a conflictsWith b
    } yield (a, b)

    if (failures.isEmpty) {
      val result = existing match {
        case reqRes: RequestResponsePact =>
          val mergedInteractions = interactions ++ otherInteractions.filterNot(interactions.contains)
          new RequestResponsePact(existing.getProvider, existing.getConsumer, JavaConversions.seqAsJavaList(mergedInteractions).asInstanceOf[java.util.List[RequestResponseInteraction]])
        case messagePact: MessagePact =>
          messagePact.mergePact(newPact)
      }
      result.sortInteractions()
      MergeSuccess(result)
    } else {
      MergeConflict(failures)
    }
  }
}

File: pact-jvm-model/src/main/kotlin/au/com/dius/pact/model/PactMerge.kt
package au.com.dius.pact.model

import mu.KLogging

data class MergeResult(val ok: Boolean, val message: String, val result: Pact? = null)

/**
 * Utility class for merging two pacts together, checking for conflicts
 */
object PactMerge : KLogging() {

  @JvmStatic
  fun merge(newPact: Pact, existing: Pact): MergeResult {
    if (!newPact.compatibleTo(existing)) {
      return MergeResult(false, "Cannot merge pacts as they are not compatible")
    }
    if (existing.interactions.isEmpty() || newPact.interactions.isEmpty()) {
      existing.mergeInteractions(newPact.interactions)
      return MergeResult(true, "", existing)
    }

    val conflicts = cartesianProduct(existing.interactions, newPact.interactions)
      .filter { it.first.conflictsWith(it.second) }
    return if (conflicts.isEmpty()) {
      existing.mergeInteractions(newPact.interactions)
      MergeResult(true, "", existing)
    } else {
      MergeResult(false, "Cannot merge pacts as there were ${conflicts.size} conflict(s) " +
        "between the interactions - ${conflicts.joinToString("\n")}")
    }
  }

  private fun cartesianProduct(list1: List<Interaction>, list2: List<Interaction>): List<Pair<Interaction, Interaction>> {
    val result = mutableListOf<Pair<Interaction, Interaction>>()
    list1.forEach { item1 ->
      list2.forEach { item2 -> result.add(item1 to item2) }
    }
    return result
  }
}
